vowpalwabbit/cb_explore_adf.cc

```diff
@@ -0,0 +1,484 @@
+
+//tau first
+#define EXPLORE_FIRST 0
+//epsilon greedy
+#define EPS_GREEDY 1
+// bagging explorer
+#define BAG_EXPLORE 2
+//softmax
+#define SOFTMAX 3
+//cover
+#define COVER 4
+
+namespace CB_EXPLORE_ADF{
+
+  struct cb_explore_adf;
+
+  void safety(v_array<float>& distribution, float min_prob);
+
+
+  struct cb_explore_adf
+  {
+
+    v_array<example*> ec_seq;
+    v_array<action_score> id_probs;
+    
+    size_t explore_type; 
+
+    size_t tau;
+    float epsilon;
+    size_t bag_size;
+    size_t cover_size;
+    float lambda;
+    
+    size_t counter;
+    bool need_to_clear;
+    vw* all;
+    LEARNER::base_learner* scorer;
+    CB::cb_class known_cost;
+    bool learn_only;
+  
+  };
+
+  template<class T> void swap(T& ele1, T& ele2)
+  { T temp = ele2;
+    ele2 = ele1;
+    ele1 = temp;
+  }
+  
+  template <bool is_learn>
+  void predict_or_learn_first(cb_explore_adf& data, base_learner& base, v_array<example*>& examples, bool isTest, bool shared)
+  { //Explore tau times, then act according to optimal.
+    
+    v_array<action_score>& preds = examples[0]->pred.a_s;
+    data.id_probs.erase();
+
+    size_t num_actions = examples.size()-1;
+    if(shared)
+      num_actions--;
+    if(preds.size() != num_actions)
+      cerr<<"Received predictions of wrong size from CB base learner\n";
+
+    if(!is_learn || !data.learn_only) {      
+      if(data.tau) {
+	float prob = 1.0/(float)num_actions;
+	for(int i = 0;i < num_actions;i++) {
+	action_score a_s;
+	a_s.action = preds[i].action;
+	a_s.score = prob;
+	data.id_probs.push_back(a_s);
+	}
+	data.tau--;
+      }
+      else {
+	for(int i = 0;i < num_actions;i++) {
+	  action_score a_s;
+	  a_s.action = preds[i].action;
+	  a_s.score = 0.;
+	  data.id_probs.push_back(a_s);
+	}
+	data.id_probs[0].score = 1.0;
+      }    
+    }
+    
+    if (is_learn && data.known_cost.probability < 1)  
+      for(example* ec : examples)
+	base.learn(*ec);
+
+    for(size_t i = 0;i < num_actions;i++) {
+      examples[0]->pred.a_s[i].score = data.id_probs[i].score;
+      examples[0]->pred.a_s[i].action = data.id_probs[i].action;
+    }
+
+  }
+
+  template <bool is_learn>
+  void predict_or_learn_greedy(cb_explore_adf& data, base_learner& base, v_array<example*>& examples, bool isTest, bool shared)
+  { //Explore uniform random an epsilon fraction of the time.
+
+    v_array<action_score>& preds = examples[0]->pred.a_s;
+    data.id_probs.erase();
+
+    size_t num_actions = examples.size()-1;
+    if(shared)
+      num_actions--;
+    if(preds.size() != num_actions)
+      cerr<<"Received predictions of wrong size from CB base learner\n";
+
+    if(!is_learn || !data.learn_only) {
+      float prob = data.epsilon/(float)num_actions;
+      for(int i = 0;i < num_actions;i++) {
+	action_score a_s;
+	a_s.action = preds[i].action;
+	a_s.score = prob;
+	data.id_probs.push_back(a_s);
+      }
```

Comment:
nit: there are some mixed use of tabs and spaces here so it's hard to read, can you run the styling scripts found in utl/new_version?